### ISSUE TITLE
About the code of large-scale datasets

### ISSUE THREAD

**Issue by ZulunZhu** [Youth-49/NPA#1]
Thanks for sharing this code. I have a question that how do you implement the experiments of large-scale datasets such as Paper100M or Mag? Can you share this part of the code?

**Comment by Youth-49** [comment]
I'm sorry for not getting back to you sooner. The original codes were on private clusters so they may not be ready for a quick start. But the implementation on paper100M or Mag is similar with small datasets. Some hints may be helpful:
1. For large graphs, we use 3-layer MLP(with hid_dim=256) as the classifier instead of a linear classifier in the original SGC.
```
class MLP(torch.nn.Module):
    def __init__(self, in_channels: int, hidden_channels: int,
                 out_channels: int, num_layers: int, dropout: float = 0.0,
                 batch_norm: bool = True, relu_last: bool = True):
        super(MLP, self).__init__()

        self.lins = ModuleList()
        self.lins.append(Linear(in_channels, hidden_channels))
        for _ in range(num_layers - 2):
            self.lins.append(Linear(hidden_channels, hidden_channels))
        self.lins.append(Linear(hidden_channels, out_channels))

        self.batch_norms = ModuleList()
        for _ in range(num_layers - 1):
            norm = BatchNorm1d(hidden_channels) if batch_norm else Identity()
            self.batch_norms.append(norm)

        self.dropout = dropout
        self.relu_last = relu_last

    def reset_parameters(self):
        for lin in self.lins:
            lin.reset_parameters()
        for batch_norm in self.batch_norms:
            batch_norm.reset_parameters()

    def forward(self, x):
        for lin, batch_norm in zip(self.lins[:-1], self.batch_norms):
            x = lin(x)
            if self.relu_last:
                x = batch_norm(x).relu_()
            else:
                x = batch_norm(x.relu_())
            x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.lins[-1](x)
        return x
```
2. The Adj we used is the symmetric normalized augmented adjacent matrix: $\hat A = \tilde{D}^{-\frac{1}{2}} (A + I)  \tilde{D}^{-\frac{1}{2}}$.
```
adj = SparseTensor(row=row, col=col, sparse_sizes=(N, N))
adj = adj.set_diag()
deg = adj.sum(dim=1).to(torch.float)
deg_inv_sqrt = deg.pow(-0.5)
deg_inv_sqrt[deg_inv_sqrt == float('inf')] = 0
adj = deg_inv_sqrt.view(-1, 1) * adj * deg_inv_sqrt.view(1, -1)
adj = adj.to_scipy(layout='csr')
```
3. we tune the lr from 5e-3 to 5e-2, and try different T (=1, =10, =50) and $\alpha$. Also, since NPA enables more propagation, >2 hop propagation would bring performance gain in many cases.
